File: src/BasicCalculator.java
/**
 Implement a basic calculator to evaluate a simple expression string.

 The expression string may contain open ( and closing parentheses ), the plus + or minus sign -, non-negative integers and empty spaces .

 You may assume that the given expression is always valid.

 Some examples:
 "1 + 1" = 2
 " 2-1 + 2 " = 3
 "(1+(4+5+2)-3)+(6+8)" = 23

 */
public class BasicCalculator {
    public int calculate(String s) {
        int o1 = 1, r1 = 0;
        // cur int
        int r2 = 0;

        Stack<Integer> stack = new Stack<>();

        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);

            if (Character.isDigit(c)) {
                int num = c - '0';

                while (i + 1 < s.length() && Character.isDigit(s.charAt(i + 1))) {
                    num = num * 10 + (s.charAt(i + 1) - '0');
                    i++;
                }

                // r2 is used to track the number;
                r2 = num;

            } else if (c == '(') {
                stack.push(o1);
                stack.push(r1);

                o1 = 1;
                r1 = 0;
            } else if (c == ')') {
                r2 = r1 + o1 * r2;

                r1 = stack.pop();
                o1 = stack.pop();
            } else if (c == '+' || c == '-') {
                r1 = r1 + o1 * r2;
                o1 = c == '+' ? 1 : -1;
                r2 = 0;
            }
        }

        return r1 + o1 * r2;
    }
}

File: Stack/calculator/basic-calculator/basic-calculator.ts
/**
Given a string s representing an expression, implement a basic calculator to evaluate it.

Example 1:

Input: s = "1 + 1"
Output: 2
Example 2:

Input: s = " 2-1 + 2 "
Output: 3
Example 3:

Input: s = "(1+(4+5+2)-3)+(6+8)"
Output: 23
 

Constraints:

1 <= s.length <= 3 * 105
s consists of digits, '+', '-', '(', ')', and ' '.
s represents a valid expression.
 */
function calculate(s: string): number {
  let stack = [];

  let result = 0;
  let number = 0;
  let sign = 1;

  for (let i = 0; i < s.length; i++) {
    const c = s[i];

    if (isDigit(c)) {
      number = 10 * number + parseInt(c);
    } else if (c === '+') {
      result += sign * number;
      number = 0;
      sign = 1;
    } else if (c === '-') {
      result += sign * number;
      number = 0;
      sign = -1;
    } else if (c === '(') {
      stack.push(result);
      stack.push(sign);
      // reset;
      sign = 1;
      result = 0;
    } else if (c === ')') {
      result += sign * number;
      number = 0;

      // sign before the parenthesis.
      result *= stack.pop();

      // result calculated before the parenthesis.
      result += stack.pop();
    }
  }

  if (number !== 0) result += sign * number;
  return result;
}

function isDigit(c: string): boolean {
  if (c === ' ') {
    return false;
  }

  return !isNaN(c as any);
}
